packages/utils/src/ts-eslint/SourceCode.ts

```diff
@@ -299,6 +299,31 @@ declare class SourceCodeBase extends TokenStore {
    * @deprecated in favor of isSpaceBetween
    */
   isSpaceBetweenTokens(first: TSESTree.Token, second: TSESTree.Token): boolean;
+  /**
+   * Returns the scope of the given node.
+   * This information can be used track references to variables.
+   * @since 8.37.0
+   */
+  getScope(node: TSESTree.Node): Scope.Scope;
+  /**
+   * Returns an array of the ancestors of the given node, starting at
+   * the root of the AST and continuing through the direct parent of the current node.
+   * This array does not include the currently-traversed node itself.
+   * @since 8.38.0
+   */
+  getAncestors(node: TSESTree.Node): TSESTree.Node[];
+  /**
+   * Returns a list of variables declared by the given node.
+   * This information can be used to track references to variables.
+   * @since 8.38.0
+   */
+  getDeclaredVariables(node: TSESTree.Node): readonly Scope.Variable[];
+  /**
+   * Marks a variable with the given name in the current scope as used.
+   * This affects the no-unused-vars rule.
+   * @since 8.39.0
+   */
+  markVariableAsUsed(name: string, node: TSESTree.Node): boolean;
```

Comment:
With these brand new methods I think it's probably best to mark them as optional so  that people have to do version checks to use them.

it's going to be a rare, rare, rare case that a plugin specifically supports `>= 8.39.0` and can only use these methods. ESLint core rules are the only case.

It's going to be an absolute dog's breakfast upgrading to support v9 when they remove the old ones :sigh: